### DISCUSSION TITLE
More than 2 states

### DISCUSSION THREAD

**Discussion by WxAaRoNxW** [vintlabs/fauxmoESP#135] [Q&A]
can this library able to have more than two states instead of just on and off?

**Comment by pvint** [comment]
The technical answer is "Yes" - you can dim lights, for example by saying "Set light to 50%". We should also be able to set colours, but it doesn't work. (I can't honestly recall if it worked in the past or not...)

**Reply by WxAaRoNxW** [reply]
How would the dimming be stated in the code?

**Reply by pvint** [reply]
Here's a `onSetState()` function I've been using for a few years (it's using ESP-IDF rather than Arduino to control the outputs, which in this case is PWM going to MOSFETs to drive bright LED strips):

```
        fauxmo.onSetState([](unsigned char device_id, const char *device_name, bool state, unsigned char val)
        {
                unsigned int v = val << 4;
                ESP_LOGI(TAG, "fauxmo device_id = %d\tstate = %d\tval = %d\tv = %d\n", device_id, state, val, v);

                if (state)
                {
                        ledc_set_fade_with_time( ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel, v, pwmFadeTime);
                        ledc_fade_start( ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel, LEDC_FADE_NO_WAIT);
                }
                else
                {
                        // turn off
                        ledc_set_duty(ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel, 0);
                        ledc_update_duty(ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel);
                }
        });
```

In my case I'm using 12 bit PWM, and the value we get from Alexa is 8 bits, so I use `v = val << 4;` to shift it to 12 bit

**Comment by barneyz** [comment]
Hello Paul,
I use my own interrupt driven fade-function for RGB and white LED strips. But the fading stutters sometimes on low brightnesses because of changing the duty cycle into an running PWM-cycle.
I test the ESP32-fuctions you use for fading with no success:
```
  ledc_set_fade_with_time( ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel, v, pwmFadeTime);
  ledc_fade_start( ledc_channel[device_id].speed_mode, ledc_channel[device_id].channel, LEDC_FADE_NO_WAIT);

```
Please can you post as an example your declaration and init sections for the used const and variables  for the arguments of both functions?

Thanks and regards
Bernd
